services/web-server/src/PulseEngine/PulseIterator.js

```diff
@@ -38,40 +41,65 @@ export default class PulseIterator {
     return this;
   }
 
+  // cause this to throw the given error for all pending and subsequent next() calls
+  pushError(error) {
+    this.stop({error, done: true});
+  }
+
+  // insert a value into the queue; the resulting promise resolves when the value
+  // is consumed, and rejects if the iterator finishes without consuming the value
   pushValue(value) {
-    if (this.pullQueue.size !== 0) {
-      this.pullQueue.first()({ value, done: false });
-      this.pullQueue = this.pullQueue.shift();
-    } else {
-      this.pushQueue = this.pushQueue.push(value);
+    if (!this.listening) {
+      return Promise.reject(new Error('iterator cancelled'));
     }
+    return new Promise((resolve, reject) => {
+      this.pushQueue = this.pushQueue.push({resolve, reject, value, done: false});
+      this.match();
+    });
   }
 
+  // pull a value from the iterator; the resulting promise resolves when a value
```

Comment:
(is `iterall` more convenient than native async iterator?)

Reply:
I'll read about `iterall` and see.. thanks!